From ac9c10d708cd65bb0e9faa62bc8847baa3f9dee3 Mon Sep 17 00:00:00 2001
From: Simon Desfarges <simonx.desfarges@intel.com>
Date: Wed, 16 Apr 2014 14:18:12 +0200
Subject: [PATCH 301/429] watchdog: check timeout values before setting it

Timeout values are checked when set, the check verifies bounds too.

Signed-off-by: Simon Desfarges <simonx.desfarges@intel.com>
---
 drivers/watchdog/intel_scu_watchdog_evo.c | 21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git a/drivers/watchdog/intel_scu_watchdog_evo.c b/drivers/watchdog/intel_scu_watchdog_evo.c
index d33933d9097d..1417f72efece 100644
--- a/drivers/watchdog/intel_scu_watchdog_evo.c
+++ b/drivers/watchdog/intel_scu_watchdog_evo.c
@@ -190,12 +190,17 @@ static void dump_softlock_debug(unsigned long data)
 #endif /* CONFIG_INTEL_SCU_SOFT_LOCKUP */
 
 /* Check current timeouts */
+/* Timeout bounds come from the MODULE_PARAM_DESC description */
 static int check_timeouts(int pre_timeout_time, int timeout_time)
 {
-	if (pre_timeout_time < timeout_time)
-		return 0;
+	if (pre_timeout_time >= timeout_time)
+		return -EINVAL;
+	if (pre_timeout_time > 160 || pre_timeout_time < 1)
+		return -EINVAL;
+	if (timeout_time > 170 || timeout_time < 1)
+		return -EINVAL;
 
-	return -EINVAL;
+	return 0;
 }
 
 /* Set the different timeouts needed by the SCU FW and start the
@@ -435,6 +440,11 @@ static long intel_scu_ioctl(struct file *file, unsigned int cmd,
 		if (get_user(val, p))
 			return -EFAULT;
 
+		if (check_timeouts(val, timeout)) {
+			pr_warn("%s: Invalid timeout thresholds (timeout: %d, pretimeout: %d) \n", __func__, timeout, val);
+			return -EINVAL;
+		}
+
 		pre_timeout = val;
 		return 0;
 	case WDIOC_SETTIMEOUT:
@@ -446,6 +456,11 @@ static long intel_scu_ioctl(struct file *file, unsigned int cmd,
 		if (get_user(val, p))
 			return -EFAULT;
 
+		if (check_timeouts(pre_timeout, val)) {
+			pr_warn("%s: Invalid timeout thresholds (timeout: %d, pretimeout: %d) \n", __func__, val, pre_timeout);
+			return -EINVAL;
+		}
+
 		timeout = val;
 		return 0;
 	case WDIOC_GETTIMEOUT:
-- 
2.37.3

